File: 프로그래머스/4/42897. 도둑질/도둑질.java
class Solution {
     public int solution(int[] money) {
        int n = money.length;
        if (n == 3) {
            return Math.max(Math.max(money[0], money[1]), money[2]);
        }

        return Math.max(rob(money, 0, n - 2), rob(money, 1, n - 1));
    }

    // 일반적인 도둑질 문제 (선형 구조)
    private int rob(int[] money, int start, int end) {
        int[] dp = new int[money.length];
        dp[start] = money[start]; // 첫번째 집만 턴 경우
        // 두번째 집은 = start만 턴경우 or stat+1만 턴 경우 
        dp[start + 1] = Math.max(money[start], money[start + 1]);
        
        for(int i = start+2; i<= end; i++){
            // 선택x, 선택o일 때
            dp[i] = Math.max(dp[i-1], dp[i-2] + money[i]);
        }
        
        return dp[end];
        
    }
}

File: 프로그래머스/4/42897. 도둑질/도둑질.py
def operate(room):
    n = len(room)
    if n == 0:
        return 0
    if n == 1:
        return room[0]
    if n == 2:
        return max(room[0], room[1])
    
    dp = [0] * n
    dp[0] = room[0]
    dp[1] = max(room[0], room[1])
    for i in range(2, n):
        dp[i] = max(dp[i-1], dp[i-2] + room[i])
    
    return dp[-1]

def solution(money):
    if len(money) == 3:
        # 3개만 있는 마을이면 바로 최댓값 리턴
        return max(money)
    
    # 4개 이상인 곳은 체크 필요
    return max(operate(money[:-1]), operate(money[1:]))